Update api endpoints

devkabiir opened the pull request:
This updates/adds api endpoints required for a working editor in flutter

- Update project apis for mutating project specific jobs/notes
- Update task->group/job api to update nested groups specific to that group/job
- Update task->group/job api to update tasks specific to that group/job
- Update task->task api to update lineitems specific to that task
- Adds search endpoints for Job/Group/Task apis

devkabiir committed:
🐛fix (project) Add missing `GET` method for `/note` endpoint

#time: 6 minutes

devkabiir committed:
🚧wip (project) Add `jobs` endpoint

devkabiir committed:
🧪🐛fix (project) Fix `test_create_job`

#time: 7 minutes

devkabiir committed:
✅test (project) Add `test_list_jobs`

devkabiir committed:
fixup! 🚧wip (project) Add `jobs` endpoint

devkabiir committed:
fixup! 🚧wip (project) Add `jobs` endpoint

devkabiir committed:
fixup! 🧪🐛fix (project) Fix `test_create_job`

devkabiir committed:
✨feat (project) Add new `job` endpoint

devkabiir committed:
♻refactor (task) Include `pk` in `Job` serialization

devkabiir committed:
🔧conf (vscode) Increase code formatting timeout for `*.py`

This might help make auto-formatting more consistent and reliable

#time: 6 minutes

devkabiir committed:
✅test (task) Add tests for `GroupSerializer` and `TaskSerializer`

devkabiir committed:
🧪🐛fix (task) Fix nested group tests

devkabiir committed:
✅test (task) Add `can_serialize_group_with_sub_group_with_no_tasks`

devkabiir committed:
✅test (task) Add more tests with lineitems and depths

devkabiir committed:
🧪🐛fix (task) Use correct expected values in asserts

devkabiir committed:
🧪🐛fix (task) Add `depth`s, use `parent` instead of `job` in `GroupFactory` & correct asserts

devkabiir committed:
✨feat (company) Allow android localhosted access with default company

This makes sure that any requests made through the api from flutter (during devlopment) are always
made to the API_DEFAULT_CMPANY and it also allows accessing the API_DEFAULT_COMPANY on the
ANDROID_LOCALHOST_LOOPBACK

#time: 2 hours 50 minutes

devkabiir committed:
🧪🐛fix (task) Correct asserts for decimals in responses

devkabiir committed:
🔧conf (vscode) Disable formatOnPaste and formatOnType

This is corrently not supported by the black formatter

devkabiir committed:
🐛fix (task) Make sure `.all()` is called for `groups` and `tasks` fields

In certain scenarios, `instance.groups.all()` should be called, but it was not being called. This
caused the serialization to fail. This is the first try at a working `TaskSerializer`,
`GroupSerializer`, as well as `JobSerializer`

#time: 1 minute

devkabiir committed:
➕dep (app) Use the official hg sources for reportlab and preppy

devkabiir commented on systori/apps/task/serializers.py:
```diff
@@ -1,58 +1,17 @@
-                else:
-                    # RecursiveField directly inside a Serializer
-                    parent_class = parent.__class__
-                proxied = parent_class(many=True, required=False)
```
Here `many=True` makes the recursive field always convert itself to `ListSerializer(child=GroupSerializer())`. But the problem is that if we use a `ListField(child=RecursiveField())` then the resulting serializer becomes `ListSerializer(child=ListSerializer(child=GroupSerializer()))` Which is wrong.
I suspect this was probably done because we only ever use `RecursiveField `in the `GroupSerializer`. And inside of that it's _always_ a list of group instances.

Anyhow, the `RecursiveField` from `djangorestframwork-recursive` works nicely

elmcrest replied:
ok perfect

devkabiir commented on systori/apps/task/serializers.py:
```diff
@@ -224,13 +236,13 @@ def perform(self):
-    groups = GroupSerializer(many=True, required=False)
-    tasks = TaskSerializer(many=True, required=False)
+    groups = GroupSerializer(required=False, many=True)
+    tasks = TaskSerializer(required=False, many=True)
```
These are unnecessary when looked at the git diff, but they happened over the course of multiple commits and experimentation, so no easy way of reversing those commits.

elmcrest replied:
lol, ok, np

devkabiir commented on systori/apps/task/serializers.py:
```diff
@@ -241,3 +253,6 @@ def update(self, job, validated_data):
         return self._data
+
+    def to_representation(self, instance):
+        return GroupSerializer.to_representation(self, instance)
```
This is temporary, by adding more tests, this will probably break things.

devkabiir committed:
♻refactor (task) Include `status`, `job`, `group` in TaskSerializer

devkabiir committed:
🐛fix (task) Manually serialize `Job`

This is to make sure `.all()` is properly called on `groups` and `tasks`

devkabiir committed:
✅test (task) Add/Update serializer tests

elmcrest commented on systori/apps/company/middleware.py:
```diff
@@ -24,6 +24,21 @@ def process_request(request):
 
         request.company = None
 
+        is_android_development = (
```
I think you should wrap this with a check if django is running in a dev environment...

something like
```python
from django.conf import settings

if settings.DEBUG:
    ....
```

devkabiir committed:
🐛fix (company) Only allow android_development in DEBUG mode

See https://github.com/systori/systori/pull/393#discussion_r329969001 for more info

devkabiir committed:
♻refactor (project) Require `HasLaborerAccess` permission for `ProjectModelViewSet`

devkabiir committed:
♻refactor (task) Exclude `status` field from `Task` serializer

This is currently not needed
This field is an `enum` with it's default set to an empty string, the
default needs to be either a valid enum value or null. Since that's how it is expected from the the
flutter client

devkabiir committed:
♻refactor (task) Include `order` field in `Job` serializer

This ia required by the flutter client in order to properly generate gaeb codes for all of the
children models like Group, Task etc

devkabiir committed:
fixup! ♻refactor (task) Exclude `status` field from `Task` serializer

devkabiir committed:
fixup! ♻refactor (task) Include `order` field in `Job` serializer

devkabiir committed:
♻refactor (task) Include `parent`, `job`, fields in `GroupSerializer`

devkabiir committed:
♻refactor (task) Include `job`, `task` fields in `LineItemSerializer`

devkabiir committed:
♻refactor (task) Add `lineitem`, `lineitems` actions in `TaskModelViewSet`

devkabiir committed:
fixup! ♻refactor (task) Add `lineitem`, `lineitems` actions in `TaskModelViewSet`

devkabiir committed:
🐛fix (task) Serialize `job` field to integer in task/lineitem serializers

devkabiir committed:
♻refactor (task) Make foreignkeys read-only in serializers

devkabiir committed:
♻refactor (task) Return full json outputs after updating Job/Group/Task/LineItem

Only `{ "pk": model.pk, "token": model.token }` was being returned as opposed to full json, which is
expected by the flutter client

devkabiir committed:
✅test (task) Add tests for `lineitem` and `lineitems` endpoint

devkabiir committed:
⏪ revert "♻refactor (task) Return full json outputs after updating Job/Group/Task/LineItem"

This reverts commit 2372c508487cb206cbcdcbddef4a384e24cc009f.

devkabiir committed:
⏪ revert "♻refactor (task) Make foreignkeys read-only in serializers"

This reverts commit 2b25f8df4ad6989c0a5bf044101c25a076c46297.

devkabiir committed:
fixup! ✨feat (task) Expose `Task` model apis

devkabiir committed:
♻refactor (task/project) Create separate serializers for apis used by flutter

The existing serializers behave in a certain way that is requried for the web editor to work. Things
such as inclusion of `token` field after serialization of Job/Group/Task/LineItem are required by
the web editor (it needs those for model uniqueness before the model has been created in the db,
i.e. no primary-key) .These things are not required by flutter client, however flutter client
expects the Serializers to be in their default behavior at most. The serializers also need to
support updates/creates for nested fields, existing serializers did support that, but they either
requried presence of `token` field or they responded with incomplete/invalid json.

devkabiir committed:
fixup! 🚧wip (project) Add `jobs` endpoint

devkabiir committed:
✨feat (task) Add `task` & `tasks` endpoint in group api

devkabiir committed:
✨feat (task) Add `task` & `tasks` endpoint in job api

devkabiir committed:
✨feat (task) Add `subgroups` endpoint in group api

devkabiir committed:
✨feat (task) Add `groups` endpoint in job api

devkabiir committed:
🐛fix (task) Set parent when creating a group

The issue with not setting the parent is that, this tries to update the group if it exists and to do
that the condition to fetch the group from db is `job=job` but that job itself has job pointing to
itself and jobs are saved in the same db table (task_group). By adding parent we make sure the
correct object is returned. Wihtout parent the job will be included as well

devkabiir committed:
🐛fix (task) Remove unused field `delete`

devkabiir committed:
🧪🐛fix (task) Fix tests failing due to 857a48df & add more tests

devkabiir committed:
♻refactor (task) Don't make unnecessary queries

devkabiir committed:
fixup! ♻refactor (task) Don't make unnecessary queries

devkabiir committed:
♻refactor (task) Raise 404 when mutating object does not exist

devkabiir committed:
✨feat (task) Add search endpoint in group api

devkabiir committed:
✨feat (task) Add search endpoint in task api

devkabiir committed:
🔧conf (vscode) Replace deprecated config keys for python extension

devkabiir committed:
🔧conf (vscode) Include 3rd party code while debugging

This allows for setting breakpoints in external dependencies or packages

devkabiir committed:
♻refactor (task) Improve generated operation id for search endpoints

devkabiir committed:
fixup! ♻refactor (task) Improve generated operation id for search endpoints

devkabiir committed:
♻refactor (task) Mark editor api for deprecation in generated schema

devkabiir committed:
♻refactor (task) Force create task in TaskSerializer.create

devkabiir committed:
♻refactor (task) Force create group in GroupSerializer.create

devkabiir committed:
♻refactor (task) Provide `precision` for `qty`,`price`,`total` in generated schema

#time: 1 minute

devkabiir committed:
♻refactor (task) Expose and require `token` field of lineitem

#time: 1 minute

devkabiir committed:
♻refactor (task) Remove endpoint operations that dont make sense

We don't allow partially updating task as well as manally creating/updating lineitems

devkabiir committed:
✨feat (task) Add validation for task and lineitems serializers

This tries to make sure invalid tasks are never allowed

devkabiir committed:
🧪♻refactor (task) Add task validation tests and fix previous tests

devkabiir committed:
✅test (task) Add tests for flutter specific serializers

devkabiir committed:
♻refactor (task) Make lineitem `token` required for create and `pk` for update

Also prevent token from being updated

devkabiir committed:
✅test (task:lineitem) Add `test_reject_update_task_lineitem_token`

devkabiir committed:
🚧wip (task:search) Only return flat objects to reduce json output

This only returns the pk, name, description and count of children objects for each search match.
This significantly reduces the json output produced

devkabiir committed:
♻refactor (task:flutter_serializers) Only check for duplicate tokens in new lineitems

devkabiir committed:
🧪🐛fix (task:api) Correct expectations for search tests

devkabiir committed:
✅test (task:api) Add `test_reject_update_task_lineitem_using_invalid_pk`

devkabiir committed:
fixup! ✨feat (task) Add search endpoint in group api

devkabiir committed:
fixup! ✨feat (task) Add search endpoint in task api

devkabiir committed:
✨feat (task:api) Add clone endpoint in group api

devkabiir committed:
✨feat (task:api) Add clone endpoint in task api

devkabiir committed:
🧪♻refactor (task:api) Reorganize task test where they make sense

devkabiir committed:
♻refactor (task:api) Pluralize search endpoint opration ids

devkabiir committed:
fixup! ♻refactor (task:api) Pluralize search endpoint opration ids

devkabiir committed:
♻refactor (task:flutter_serializers) Disallow updating lineitems from other tasks/jobs

devkabiir committed:
♻refactor (task:flutter_serializers) Prevent task creation when lineitem has a pk

devkabiir committed:
♻refactor (task:flutter_serializers) Delete absent lineitems on task create/update

This removes the need to separately make an api request to delete lineitems by the client

devkabiir committed:
♻refactor (task:flutter_serializers) Prevent task update for duplicate lineitem pk

devkabiir committed:
✅test (task:api) Add `test_reject_create_task_with_lineitem_with_pk`

devkabiir committed:
fixup! 🧪♻refactor (task:api) Reorganize task test where they make sense

devkabiir committed:
fixup! ♻refactor (task:flutter_serializers) Delete absent lineitems on task create/update

devkabiir committed:
fixup! ♻refactor (task:flutter_serializers) Prevent task update for duplicate lineitem pk

devkabiir committed:
fixup! 🧪♻refactor (task:api) Reorganize task test where they make sense

devkabiir committed:
♻refactor (task:api) Remove unused endpoints for direct task manipulation

Use job/group specific api endpoints for manipulating their tasks

devkabiir committed:
♻refactor (task:api) Allow reading arbitrary tasks

This is required to fetch selected task during task search or auto complete

devkabiir committed:
♻refactor (task) Add project and job name in group search result

This helps add more context

devkabiir committed:
fixup! ♻refactor (task) Add project and job name in group search result

devkabiir committed:
♻refactor (task) Add project and job name in task search result

This helps add more context

devkabiir committed:
🐛fix (task:flutter_serializers) Consider hidden lineitems when validating task total

#time: 10 minutes

devkabiir committed:
🐛fix (task:flutter_serializers) Round expected total to 2 before compare

#time: 3 minutes

devkabiir committed:
🐛fix (task:flutter_serializers) Allow null qty for t&m tasks

devkabiir committed:
🐛fix (task:flutter_serializers) Take into account that any cell may be calculated

Previously if price was a calculated value from qty and total. The validation would fail even if all
cells had correct values. This is due to the rounding error when checking the calculated value
against expected value.
